p2p/p2pimpl/p2pimpl.go

```diff
@@ -1,25 +1,57 @@
 package p2pimpl
 
 import (
+	"fmt"
+	"net"
+
 	"github.com/harmony-one/harmony/p2p"
 	"github.com/harmony-one/harmony/p2p/host"
 	"github.com/harmony-one/harmony/p2p/host/hostv1"
 	"github.com/harmony-one/harmony/p2p/host/hostv2"
+
+	"github.com/harmony-one/harmony/internal/utils"
+	peer "github.com/libp2p/go-libp2p-peer"
+	multiaddr "github.com/multiformats/go-multiaddr"
 )
 
 // Version The version number of p2p library
 // 1 - Direct socket connection
 // 2 - libp2p
 const Version = 2
 
-// NewHost starts the host
-func NewHost(peer p2p.Peer) host.Host {
-	// log.Debug("New Host", "ip/port", net.JoinHostPort(peer.IP, peer.Port))
+// NewHost starts the host for p2p
+// for hostv2, it generates multiaddress, keypair and add PeerID to peer, add priKey to host
+// TODO (leo) the PriKey of the host has to be persistent in disk, so that we don't need to regenerate it
+// on the same host if the node software restarted. The peerstore has to be persistent as well.
+func NewHost(self *p2p.Peer) (host.Host, error) {
 	if Version == 1 {
-		h := hostv1.New(peer)
-		return h
+		h := hostv1.New(self)
+		return h, nil
+	}
+
+	selfAddr, err := multiaddr.NewMultiaddr(fmt.Sprintf("/ip4/0.0.0.0/tcp/%s", self.Port))
```

Comment:
Just a note for future that the port may be an ephemeral port later when we have node discovery, which is a node ID → node locator (such as a multiaddr) lookup service.  In that case, a node would just let the OS pick an unused port (in the socket interface one does this by saying port 0; unsure what the libp2p way is) and bind to it, then later register its node-to-locator association with the node discovery/rendezvous service, using its own IP address + the ephemeral port number.

(This logic gets further complicated when we consider NAT traversal, but let's not go too deep :-p)

Reply:
@harmony-ek , please open an issue for further note.